### PR TITLE
Option for covering tests

### LINKED ISSUE #96: Option for intrumenting and computing the coverage of test classes
Right now, only source classes are instrumented. We should have an option for enabling test instrumentation too.

### PR DESCRIPTION
**Pull request by andre15silva**
This PR presents a new option for test-runner, covering test classes.

Closes #96 
Related to https://github.com/SpoonLabs/flacoco/issues/41

### CODE REVIEW

File: src/main/java/eu/stamp_project/testrunner/runner/coverage/JUnit4JacocoRunnerCoveredResultPerTestMethod.java [hunk 1]
```diff
@@ -25,10 +25,12 @@ public JUnit4JacocoRunnerCoveredResultPerTestMethod(String classesDirectory, Str
 
 	public JUnit4JacocoRunnerCoveredResultPerTestMethod(String classesDirectory, String testClassesDirectory, List<String> blackList, CoverageTransformer coverageTransformer) {
 		super(classesDirectory, testClassesDirectory, blackList, coverageTransformer);
+		System.out.println("I'M EVERYWHERE : " + classesDirectory);
```
Comment (danglotb) [inline src/main/java/eu/stamp_project/testrunner/runner/coverage/JUnit4JacocoRunnerCoveredResultPerTestMethod.java:28]: Could you remove these sout (for debugging I guess)? :feelsgood:

File: src/main/java/eu/stamp_project/testrunner/runner/coverage/JUnit4JacocoRunnerCoveredResultPerTestMethod.java [hunk 2]
```diff
@@ -25,10 +25,12 @@ public JUnit4JacocoRunnerCoveredResultPerTestMethod(String classesDirectory, Str
...
 	}
 
 	@Override
 	protected CoveredTestResultPerTestMethod executeCoveredTestPerTestMethod(RuntimeData data, String classesDirectory, String[] testClassNames, String[] testMethodNames) {
+		System.out.println("I'M HERE : " + classesDirectory);
```
Comment (danglotb) [inline src/main/java/eu/stamp_project/testrunner/runner/coverage/JUnit4JacocoRunnerCoveredResultPerTestMethod.java:33]: Same here

File: src/main/java/eu/stamp_project/testrunner/runner/coverage/JUnit4JacocoRunnerCoveredResultPerTestMethod.java (hunk 1 above)
Comment (andre15silva) [reply src/main/java/eu/stamp_project/testrunner/runner/coverage/JUnit4JacocoRunnerCoveredResultPerTestMethod.java:28]: Ups, forgot I had those there :grimacing:

**Review by danglotb — APPROVED** [review]

**Comment by andre15silva** [comment]
Btw, when using this in `flacoco` I get an issue only when the `coverTests` flag is true

```
Exception in thread "main" java.lang.RuntimeException: fr/spoonlabs/FLtest1/Calculator.class,/tmp/junit4526515788628245870/fr/spoonlabs/FLtest1/Calculator.class,fr.spoonlabs.FLtest1.Calculator
	at eu.stamp_project.testrunner.runner.coverage.JacocoRunner.instrumentAll(JacocoRunner.java:301)
	at eu.stamp_project.testrunner.runner.coverage.JacocoRunner.<init>(JacocoRunner.java:84)
	at eu.stamp_project.testrunner.runner.coverage.JacocoRunnerPerTestMethod.<init>(JacocoRunnerPerTestMethod.java:99)
	at eu.stamp_project.testrunner.runner.coverage.JacocoRunnerCoveredResultPerTestMethod.<init>(JacocoRunnerCoveredResultPerTestMethod.java:34)
	at eu.stamp_project.testrunner.runner.coverage.JUnit4JacocoRunnerCoveredResultPerTestMethod.<init>(JUnit4JacocoRunnerCoveredResultPerTestMethod.java:27)
	at eu.stamp_project.testrunner.runner.coverage.JUnit4JacocoRunnerCoveredResultPerTestMethod.main(JUnit4JacocoRunnerCoveredResultPerTestMethod.java:54)
Caused by: java.io.IOException: Error while instrumenting class fr.spoonlabs.FLtest1.Calculator.
	at org.jacoco.core.instr.Instrumenter.instrumentError(Instrumenter.java:166)
	at org.jacoco.core.instr.Instrumenter.instrument(Instrumenter.java:138)
	at eu.stamp_project.testrunner.runner.coverage.JacocoRunner.instrumentAll(JacocoRunner.java:299)
	... 5 more
Caused by: java.lang.IllegalArgumentException
	at org.jacoco.core.internal.Java9Support.readFully(Java9Support.java:45)
	at org.jacoco.core.instr.Instrumenter.instrument(Instrumenter.java:136)
	... 6 more

```

**Comment by danglotb** [comment]
Hi @andre15silva 

I'm sorry, but I have no clue on what happening here.

> It happens when instrumenting the class files for normal source code classes, which come from the same place as the normal tests.
Could you explain a bit more? I'm sorry, I did not fully understood

**Comment by andre15silva** [comment]
> Could you explain a bit more? I'm sorry, I did not fully understand

The last commit fixes the issue. The classloader in `JacocoRunner` wasn't being initialized according to the new input format.

Don't merge just yet tho, the coverage is including the tests but then doesn't register which instructions are ran.

**Comment by andre15silva** [comment]
So, I think there is an issue with the way `test-runner` uses the JUnit listeners.

I have added two test cases that showcase the difference between the old way `flacoco` was using[1], and the new one through the `EntryPoint`[2].

[1] The old way of `flacoco` using `test-runner` is here https://github.com/SpoonLabs/flacoco/blob/4a99453ffc01305741b946049c3f7edef29e26b6/src/main/java/fr/spoonlabs/flacoco/core/coverage/CoverageRunner.java#L42-L72.

The corresponding function in `JacocoRunner` is https://github.com/STAMP-project/test-runner/blob/d7dd72692ffdafb986e5fa53c2bdc2a40e75ce56/src/main/java/eu/stamp_project/testrunner/runner/coverage/JacocoRunner.java#L233-L285.

[2] The new way of using `test-runner`, through the `EntryPoint` means that we have a listener (a `RunListener` in the case of JUnit4), which does the initialization and collection previously done in the `run` method of `JacocoRunner` when the tests start and finish https://github.com/STAMP-project/test-runner/blob/d7dd72692ffdafb986e5fa53c2bdc2a40e75ce56/src/main/java/eu/stamp_project/testrunner/listener/junit4/CoveredTestResultsPerJUnit4TestMethod.java#L63-L92

So, after debugging, my current hypothesis is that the issue is related to this difference, with the test statements somehow not being recorded until the whole execution finishes? 
I have modifying this, but didn't get anywhere. I have searched for similar issues, but haven't found any.

@danglotb @martinezmatias do any of you have any idea what could be causing this?

**Comment by danglotb** [comment]
I did not completely understand the issue but one thing is certain:

You should not use directly the `JacocoRunner`, as it would result in a conflict of classloading, but rather use the API provided by `EntryPoint`, which will run a `java` command line, and therefore starts a new JVM, avoiding any classloading problems.

**Comment by andre15silva** [comment]
> You should not use directly the `JacocoRunner`, as it would result in a conflict of classloading, but rather use the API provided by `EntryPoint`, which will run a `java` command line, and therefore starts a new JVM, avoiding any classloading problems.

Yes, definitely. That's why we already moved to using `EntryPoint` (see https://github.com/SpoonLabs/flacoco/pull/35)

> I did not completely understand the issue

So, one feature that was broken when we started using `EntryPoint` was the coverage of test classes being computed, which this PR is meant to fix. The instrumentation works, but the returned results are all zeros.

For example:
- Obtained through `JacocoRunner`: `example/TestSuiteExample=Lines covered [init=2, end=45, cov={2=3, 3=0, 5=0, 6=0, 7=0, 8=0, 9=0, 12=0, 13=0, 14=0, 15=0, 16=0, 19=0, 20=0, 21=0, 22=0, 25=4, 26=7, 27=1, 28=0, 31=0, 32=0, 33=0, 34=0, 37=0, 38=0, 39=0, 40=0, 43=0, 44=0, 45=0, 46=0}]`
- Obtained through `EntryPoint`: `example/TestSuiteExample=Lines covered [init=2, end=45, cov={2=0, 3=0, 5=0, 6=0, 7=0, 8=0, 9=0, 12=0, 13=0, 14=0, 15=0, 16=0, 19=0, 20=0, 21=0, 22=0, 25=0, 26=0, 27=0, 28=0, 31=0, 32=0, 33=0, 34=0, 37=0, 38=0, 39=0, 40=0, 43=0, 44=0, 45=0, 46=0}]`

My hypothesis is that this occurs due to the data collection having moved from the `run` method we were using in `JacocoRunner` to the JUnit listeners. Is it possible that the instructions of test classes are only registered after the listener method `testFinished` has been called?

**Comment by danglotb** [comment]
I'm sorry I cannot see it through right now but I remember having such troubles by the time I was developing  `test-runner`.

I fetch your branch and ran `eu.stamp_project.testrunner.runner.coverage.JacocoRunnerTest#testRunCoveredTestResults` and obtained: 

```txt
java.lang.AssertionError: 
Expected :Optional[2]
Actual   :3
```

I'm sorry but I would like to have a complete scenario that allows me to reproduce. Could you provide me this or help me to find it?

**Comment by andre15silva** [comment]
I'm so sorry, I seem to have reverted a change before committing :facepalm: . Assertions are working now.

The two tests in question are `EntryPointTest#testRunCoveredTestResultPerTestMethodsDetailedCoverageCoverTests` and `JacocoRunnerTest#testRunCoveredTestResults`.

The first one shows the buggy behavior, since the expected result is the one the second one returns. I added some prints in the test cases to show the two versions that should be equal.

This means that the coverage of class `example.TestSuiteExample`, of which method `test8` is ran, should not be full of 0s, but have some execution going through it.

I think these two tests provide the complete scenario, with the expected behavior from the previous version, and the buggy behavior from the new version, but let me know if you need anything else! Thank you for your time :)

**Comment by danglotb** [comment]
Hi @andre15silva 

Is this still a work in progress or should we close it?

**Comment by andre15silva** [comment]
> Is this still a work in progress or should we close it?

Hi @danglotb. I'd say we close it, I'm no longer working on it.

**Pull request closed without merging**